drivers/hwmon/pmbus/pmbus_core.c

```
@@ -186,6 +189,98 @@ int pmbus_set_page(struct i2c_client *client, int page)
 }
 EXPORT_SYMBOL_GPL(pmbus_set_page);
 
+static void i2c_pmbus_try_get_dmabuf(struct i2c_msg *msg)
+{
+	unsigned char *dma_buf;
+
+	dma_buf = kzalloc(PMBUS_BLOCK_MAX + 3, GFP_KERNEL);
+	if (!dma_buf)
+		return;
+
+	msg->buf = dma_buf;
+	msg->flags |= I2C_M_DMA_SAFE;
+}
+
+/* Block Write/Read command.
+ * @client: Handle to slave device
+ * @cmd: Byte interpreted by slave
+ * @w_len: Size of write data block; PMBus allows at most 255 bytes
+ * @data_w: byte array which will be written.
+ * @data_r: Byte array into which data will be read; big enough to hold
+ *	the data returned by the slave. PMBus allows at most 255 bytes.
+ *
+ * Different from Block Read as it sends data and waits for the slave to
+ * return a value dependent on that data. The protocol is simply a Write Block
+ * followed by a Read Block without the Read-Block command field and the
+ * Write-Block STOP bit.
+ *
+ * Returns number of bytes read or negative errno.
+ */
+int pmbus_block_wr(struct i2c_client *client, u8 cmd, u8 w_len,
+		   u8 *data_w, u8 *data_r)
+{
+	u8 write_buf[PMBUS_BLOCK_MAX + 1];
+	u8 read_buf[PMBUS_BLOCK_MAX + 1];
+	struct i2c_msg msgs[2] = {
+		{
+			.addr = client->addr,
+			.flags = 0,
+			.buf = write_buf,
+			.len = w_len + 2,
+		},
+		{
+			.addr = client->addr,
+			.flags = I2C_M_RD,
+			.buf = read_buf,
+			.len = PMBUS_BLOCK_MAX,
+		}
+	};
+	u8 addr = 0;
+	u8 crc = 0;
+	int ret;
+
+	i2c_pmbus_try_get_dmabuf(&msgs[0]);
```

i'm confused;
`struct i2c_msg msgs[2]` has some buffers attached to it, from the stack, but then this comes and allocs some new buffers?

---

This method is taken from the SmBus Linux implementation.

---

Do we want to unconditional use dma safe buffer here? How bit are typically these messages?
And why not using i2c apiS `i2c_get_dma_safe_msg_buf()` and `i2c_put_dma_safe_msg_buf`?

---

Most messages will be of 255 length bytes (when writing fw/configs or reading blackboxes).
ADM1266 data-sheet (pag. 55): "Read back user data length, maximum value of N = 252".

---

Ok, then makes sense to me to always go with dma safe buffers. Still, Why not using the i2c functions?  If you stick with `i2c_pmbus_try_get_dmabuf`, I would return an error and make the use of dma safe buffer mandatory (meaning, just bail out if you cannot allocate a buffer)...